### PR TITLE
Adding receiver emitter nodes

### PR DESCRIPTION
**Pull request by angel-ayala**
Hi,

I'm currently following the [Tutorial Creating a Custom Cpp Plugin](https://github.com/cyberbotics/webots_ros2/wiki/Tutorial-Creating-a-Custom-Cpp-Plugin) to include the [Emitter](https://www.cyberbotics.com/doc/reference/emitter?tab-language=c) and [Receiver](https://www.cyberbotics.com/doc/reference/receiver?tab-language=c). nodes.

This added:
- Two plugins to `webots_ros2_driver` module: `Ros2Emitter` and `Ros2Receiver`.
- One `StringStamped` message and `SetString` service to `webots_ros2_msgs`.

Current dev-stage allows:
-  the `Receiver` node publish in a `StringStamped` topic  `/data` if `wb_receiver_get_queue_length(mReceiver) > 0` the results of `wb_receiver_get_data(mReceiver)` .
- the `Emitter` node create a `SetString` service `/send` to perform the `wb_emitter_send()` function.
Stills require to implement all the remaining functions.

**TODO**
  - [ ] `Emitter` node's missing functions.
  - [ ] `Receiver` node's missing functions.
  - [ ] perform tests.

**Additional context**
I'm currently developing another package in order to use these plugins simulating a communication between two radio frequency controllers.
I have a `RCController` which controls a `RadioController` node. 
There are two `RadioController` nodes in the scene: `RCControlMain` and `RCControlSlave`

When I launch the `robot_launch.py` file, seems that the  `RCControlMain` keeps waiting for the `/send` service to start.
The expected behavior is a continuous string sharing between both `RadioController`, but seems that one is being replaced by the other.

Any thoughts?

Here is the console's output:
```
$ ros2 launch rc_comm robot_launch.py 
[INFO] [launch]: All log files can be found below /home/kelo/.ros/log/2023-03-03-15-25-53-387733-mjro-linux-hp-66077
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [webots-1]: process started with pid [66078]
[INFO] [ros2_supervisor.py-2]: process started with pid [66080]
[INFO] [driver-3]: process started with pid [66082]
[INFO] [driver-4]: process started with pid [66085]
[driver-3] Cannot connect to Webots instance, retrying for another 50 seconds...
[driver-4] Cannot connect to Webots instance, retrying for another 50 seconds...
[webots-1] Warning: QApplication: invalid style override 'KvAdaptaDark' passed, ignoring it.
[webots-1] 	Available styles: Windows, Fusion
[ros2_supervisor.py-2] Cannot connect to Webots instance on socket "/tmp/webots-1234/ipc/Ros2Supervisor/extern", retrying for another 50 seconds...
[driver-4] [INFO] [1677867957.553540411] [ActionReceiver]:  initialized!
[driver-3] [INFO] [1677867957.554331172] [StateReceiver]:  initialized!
[driver-3] [INFO] [1677867957.554829176] [ActionEmitter]: Emitter initialized!
[driver-4] [INFO] [1677867957.555852534] [StateEmitter]: Emitter initialized!
[driver-4] [INFO] [1677867957.776921623] [radio_controller]: RCControlSlave with ActionReceiver and StateEmitter initiated
[driver-3] [INFO] [1677867957.780533304] [radio_controller]: RCControlMain with StateReceiver and ActionEmitter initiated
[driver-4] [INFO] [1677867958.794008276] [radio_controller]: ActionEmitter's still not available, waiting 1 seconds...
[driver-3] [INFO] [1677867958.796117445] [radio_controller]: ActionEmitter's still not available, waiting 1 seconds...
[driver-4] [INFO] [1677867960.797089514] [radio_controller]: ActionEmitter's still not available, waiting 2 seconds...
[driver-3] [INFO] [1677867960.799117403] [radio_controller]: ActionEmitter's still not available, waiting 2 seconds...
[driver-4] [INFO] [1677867963.801271317] [radio_controller]: ActionEmitter's still not available, waiting 3 seconds...
[driver-3] [INFO] [1677867963.803361786] [radio_controller]: ActionEmitter's still not available, waiting 3 seconds...
```

### Plugin File
`rc.py`
```python
import rclpy
import json

from webots_ros2_msgs.srv import SetString
from webots_ros2_msgs.msg import StringStamped

class RCController:
    def init(self, webots_node, properties):
        # Runtime vars
        self.__time_counter = 0
        self.__lift_thrust = 68.5  # with this thrust, the drone lifts.
        self.__robot = webots_node.robot
        self.__timestep = self.__robot.getBasicTimeStep()

        # ROS interface
        rclpy.init(args=None)
        self.__node = rclpy.create_node('radio_controller')

        # Initialize comms
        self.__rx_state = dict()
        
        emitter_name = 'ActionEmitter' if 'Main' in self.__robot.getName() else 'StateEmitter'
        receiver_name = 'StateReceiver' if 'Main' in self.__robot.getName() else 'ActionReceiver'

        self.__action_client = self.__node.create_client(
            SetString, f"/{self.__robot.getName()}/{emitter_name}/send")

        self.__node.create_subscription(
            StringStamped, f"/{self.__robot.getName()}/{receiver_name}/data",
            self.__call_rx_data, 1)

        timeout = 1
        while not self.__action_client.wait_for_service(timeout_sec=timeout):
            self.__node.get_logger().info(
                f"ActionEmitter's still not available, waiting {timeout} seconds...")
            timeout += 1

        self.__node.get_logger().info(f"{self.__robot.getName()} with {receiver_name} and {emitter_name} initiated")

    def __call_rx_data(self, data):
        self.__rx_data = data
        # read current command
        self.__node.get_logger().info(str(self.__rx_data))
        try:
            self.__rx_state = json.loads(self.__rx_data.data)

        except json.JSONDecodeError as e:
            self.__node.get_logger().error(str(e))
        finally:
            self.__node.get_logger().info("RC publishing")
            self.__node.get_logger().info(str(self.__rx_data))

    def __tx_data(self, data):
        msg = json.dumps(data)

        request = SetString.Request()
        request.value = msg

        self.__node.get_logger().info("Sending data: " + msg[:100] + '...')

        self.__node.future = self.__action_client.call_async(request)
        rclpy.spin_until_future_complete(self.__node, self.__node.future)

        return self.__node.future.result()

    def step(self):
        rclpy.spin_once(self.__node, timeout_sec=0)
        self.__tx_data('testing string')

```

### The launch file
`robot_launch.py`
```python
#!/usr/bin/env python

"""Launch Simple RC Emitter/Receiver test
"""

import os
import pathlib
import launch
from launch.substitutions import LaunchConfiguration
from launch.actions import DeclareLaunchArgument
from launch.substitutions.path_join_substitution import PathJoinSubstitution
from launch_ros.actions import Node
from launch import LaunchDescription
from ament_index_python.packages import get_package_share_directory
from webots_ros2_driver.webots_launcher import WebotsLauncher
from webots_ros2_driver.utils import controller_url_prefix

PACKAGE_NAME = 'rc_comm'


def get_ros2_nodes(*args):
    package_dir = get_package_share_directory(PACKAGE_NAME)
    rc_main_description = pathlib.Path(
        os.path.join(package_dir, 'resource', 'rc_main.urdf')).read_text()
    rc_slave_description = pathlib.Path(
        os.path.join(package_dir, 'resource', 'rc_slave.urdf')).read_text()

    rc_main_name = 'RCControlMain'
    rc_slave_name = 'RCControlSlave'

    rc_main = Node(
        package='webots_ros2_driver',
        executable='driver',
        output='screen',
        additional_env={
            'WEBOTS_CONTROLLER_URL': controller_url_prefix() + rc_main_name
            },
        parameters=[
            {'robot_description': rc_main_description},
        ]
    )

    rc_slave = Node(
        package='webots_ros2_driver',
        executable='driver',
        output='screen',
        additional_env={
            'WEBOTS_CONTROLLER_URL': controller_url_prefix() + rc_slave_name
            },
        parameters=[
            {'robot_description': rc_main_description},
        ]
    )

    return [
        rc_main,
        rc_slave,
    ]


def generate_launch_description():
    package_dir = get_package_share_directory(PACKAGE_NAME)
    world = LaunchConfiguration('world')

    webots = WebotsLauncher(
        world=PathJoinSubstitution([package_dir, 'worlds', world]),
        ros2_supervisor=True
    )

    # The following line is important!
    # This event handler respawns the ROS 2 nodes on simulation reset (supervisor process ends).
    reset_handler = launch.actions.RegisterEventHandler(
        event_handler=launch.event_handlers.OnProcessExit(
            target_action=webots._supervisor,
            on_exit=get_ros2_nodes,
        )
    )

    return LaunchDescription([
        DeclareLaunchArgument(
            'world',
            default_value='simple_rc_world.wbt',
            description=f"Choose one of the world files from `/{PACKAGE_NAME}/worlds` directory"
        ),
        webots,
        webots._supervisor,

        # This action will kill all nodes once the Webots simulation has exited
        launch.actions.RegisterEventHandler(
            event_handler=launch.event_handlers.OnProcessExit(
                target_action=webots,
                on_exit=[
                    launch.actions.UnregisterEventHandler(
                        event_handler=reset_handler.event_handler
                    ),
                    launch.actions.EmitEvent(event=launch.events.Shutdown())
                ],
            )
        ),

        # Add the reset event handler
        reset_handler
    ] + get_ros2_nodes())

```

### CODE REVIEW

**Comment by angel-ayala** [comment]
I found that the node name was settled incorrectly in the plugin file `rc.py` and updated the following lines:
```python
node_name = 'radio_controller' + ('_main' if 'Main' in self.__robot.getName() else '_slave' )
self.__node = rclpy.create_node(node_name)
```

but the same issue appears
```
$ ros2 launch rc_comm robot_launch.py
[INFO] [launch]: All log files can be found below /home/kelo/.ros/log/2023-03-04-16-49-10-327859-mjro-linux-hp-8669
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [webots-1]: process started with pid [8670]
[INFO] [ros2_supervisor.py-2]: process started with pid [8672]
[INFO] [driver-3]: process started with pid [8674]
[INFO] [driver-4]: process started with pid [8676]
[driver-3] Cannot connect to Webots instance, retrying for another 50 seconds...
[driver-4] Cannot connect to Webots instance, retrying for another 50 seconds...
[webots-1] Warning: QApplication: invalid style override 'KvAdaptaDark' passed, ignoring it.
[webots-1] 	Available styles: Windows, Fusion
[ros2_supervisor.py-2] Cannot connect to Webots instance on socket "/tmp/webots-1234/ipc/Ros2Supervisor/extern", retrying for another 50 seconds...
[driver-3] [INFO] [1677959353.483491542] [StateReceiver]: StateReceiver initialized!
[driver-3] [INFO] [1677959353.484710986] [ActionEmitter]: Emitter initialized!
[driver-4] [INFO] [1677959353.485622702] [ActionReceiver]: ActionReceiver initialized!
[driver-4] [INFO] [1677959353.486467474] [StateEmitter]: Emitter initialized!
[driver-3] [INFO] [1677959353.667974603] [radio_controller_main]: RCControlMain with StateReceiver and ActionEmitter initiated
[driver-4] [INFO] [1677959353.668868325] [radio_controller_slave]: RCControlSlave with ActionReceiver and StateEmitter initiated
[driver-3] [INFO] [1677959354.679176711] [radio_controller_main]: ActionEmitter's still not available, waiting 1 seconds...
[driver-4] [INFO] [1677959354.681989234] [radio_controller_slave]: ActionEmitter's still not available, waiting 1 seconds...
[driver-3] [INFO] [1677959356.682764096] [radio_controller_main]: ActionEmitter's still not available, waiting 2 seconds...
[driver-4] [INFO] [1677959356.685567244] [radio_controller_slave]: ActionEmitter's still not available, waiting 2 seconds...
[driver-3] [INFO] [1677959359.687935513] [radio_controller_main]: ActionEmitter's still not available, waiting 3 seconds...
[driver-4] [INFO] [1677959359.690402401] [radio_controller_slave]: ActionEmitter's still not available, waiting 3 seconds...
[driver-3] [INFO] [1677959363.694794663] [radio_controller_main]: ActionEmitter's still not available, waiting 4 seconds...
[driver-4] [INFO] [1677959363.697368319] [radio_controller_slave]: ActionEmitter's still not available, waiting 4 seconds...
```

**Comment by angel-ayala** [comment]
I noticed other error in the `robot_launch.py` file when declaring the `radio_controller_slave` node, updated as follows:
```python
    rc_slave = Node(
        package='webots_ros2_driver',
        executable='driver',
        output='screen',
        additional_env={
            'WEBOTS_CONTROLLER_URL': controller_url_prefix() + rc_slave_name
            },
        parameters=[
            {'robot_description': rc_slave_description},
        ]
    )
```

However, still keeps waiting for the `/Emitter/send` service to start.
Here is the log output:
```
$ ros2 launch rc_comm robot_launch.py
[INFO] [launch]: All log files can be found below /home/kelo/.ros/log/2023-03-04-17-54-14-392640-mjro-linux-hp-15136
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [webots-1]: process started with pid [15138]
[INFO] [ros2_supervisor.py-2]: process started with pid [15140]
[INFO] [driver-3]: process started with pid [15142]
[driver-3] Cannot connect to Webots instance on socket "/tmp/webots-1234/ipc/RCControlMain/extern", retrying for another 50 seconds...
[webots-1] Warning: QApplication: invalid style override 'KvAdaptaDark' passed, ignoring it.
[webots-1] 	Available styles: Windows, Fusion
[ros2_supervisor.py-2] Cannot connect to Webots instance on socket "/tmp/webots-1234/ipc/Ros2Supervisor/extern", retrying for another 50 seconds...
[driver-3] [INFO] [1677963257.525484012] [StateReceiver]: StateReceiver initialized!
[driver-3] [INFO] [1677963257.525892615] [ActionEmitter]: Emitter initialized!
[driver-3] [INFO] [1677963257.690790660] [radio_controller_main]: RCControlMain with StateReceiver and ActionEmitter initiated
[driver-3] [INFO] [1677963258.700027184] [radio_controller_main]: /RCControlMain/ActionEmitter/send ActionEmitter's still not available, waiting 1 seconds...
[driver-3] [INFO] [1677963260.703371037] [radio_controller_main]: /RCControlMain/ActionEmitter/send ActionEmitter's still not available, waiting 2 seconds...
[driver-3] [INFO] [1677963263.708223966] [radio_controller_main]: /RCControlMain/ActionEmitter/send ActionEmitter's still not available, waiting 3 seconds...
```

Apparently the error is in the service status of the Emitter node, but the code still compiling.
Do you have any thoughts?

**Comment by angel-ayala** [comment]
Some bugs and data type when `wb_emitter_send_data()` and `wb_receiver_get_data()` are called.

I updated the plugin file `rc.py` with a workaround for the deadlock when the service is called and seems that the added `Ros2Emitter` and `Ros2Receiver` plugins are working.
[Previous mentioned TODO tasks](https://github.com/cyberbotics/webots_ros2/pull/656#issue-1609111497) still require to be implemented

I don't know if it is the correct way, but it is useful.

Updated `rc.py`:
```python
import rclpy
import json
from webots_ros2_msgs.srv import SetString
from webots_ros2_msgs.msg import StringStamped


class RCController:

    def init(self, webots_node, properties):
        # Runtime vars
        self.__time_counter = 0
        self.__lift_thrust = 68.5  # with this thrust, the drone lifts.
        self.__robot = webots_node.robot
        self.__timestep = self.__robot.getBasicTimeStep()

        # ROS interface
        rclpy.init(args=None)
        node_name = 'radio_controller' + ('_main' if 'Main' in 
                                          self.__robot.getName() else '_slave' )
        self.__node = rclpy.create_node(node_name)

        # Initialize comms
        self.__rx_state = dict()
        
        emitter_name = 'ActionEmitter' if 'Main' in self.__robot.getName() else 'StateEmitter'
        receiver_name = 'StateReceiver' if 'Main' in self.__robot.getName() else 'ActionReceiver'

        self.__topic = f"/{self.__robot.getName()}/{emitter_name}/send"
        self.__action_client = self.__node.create_client(
            SetString, self.__topic)

        self.__node.get_logger().info(f"{self.__robot.getName()} with {receiver_name} and {emitter_name} initiated")
        self.__node.create_subscription(
            StringStamped, f"/{self.__robot.getName()}/{receiver_name}/data",
            self.__call_rx_data, 1)

        if not self.__action_client.service_is_ready():
            timeout = 50
            self.__action_client.wait_for_service(timeout_sec=timeout)
            self.__node.get_logger().info(
                f"{self.__topic} {emitter_name}'s still not available, waiting {timeout} seconds...")

        self.__lock = False
        self.__node.get_logger().info(f"{self.__robot.getName()} with {receiver_name} and {emitter_name} initiated")

    def __call_rx_data(self, data):
        self.__rx_data = data
        # read current command
        try:
            self.__node.get_logger().info("__rx_data published")
            self.__node.get_logger().info(str(self.__rx_data))
            self.__rx_state = json.loads(self.__rx_data.data)

        except json.JSONDecodeError as e:
            self.__node.get_logger().error(str(e))
        finally:
            self.__node.get_logger().info("__rx_state")
            self.__node.get_logger().info(str(self.__rx_state))

    def __tx_data(self, data):
        # Check for future.done()
        if self.__lock:
            self.__node.get_logger().info('checking future ' + 
                                          str(self.__node.future.result()))
            if self.__node.future.result():
                self.__lock = not self.__node.future.result().success
        # else:
        if not self.__lock:
            request = SetString.Request()
            request.value = json.dumps(data)

            self.__node.get_logger().info(str(self.__action_client.service_is_ready()))
            self.__node.get_logger().info("Sending data: " + request.value[:100] + '...')
    
            self.__node.future = self.__action_client.call_async(request)
            self.__lock = True

    def step(self):
        rclpy.spin_once(self.__node, timeout_sec=0)        
        data = dict(msg=f"{self.__robot.getName()} test send")
        self.__tx_data(data)

```

**Comment by omichel** [comment]
I fixed a number of issue with the compliance with clang format in the source code and updated the code with the upstream branch. However, there are still some TODOs to address, so I am setting this PR to draft for now. Please set it ready for review once the TODOs are addressed.

**Comment by angel-ayala** [comment]
The missing function for both Emitter and Receiver node were implemented, same tests were addressed in simulated scene with successful results.

**TODO**
- [x] Emitter node's missing functions.
- [x] Receiver node's missing functions.
- [x] perform tests.

File: webots_ros2_driver/include/webots_ros2_driver/plugins/static/Ros2Emitter.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#include <webots_ros2_msgs/srv/get_float.hpp>
+#include <webots_ros2_msgs/srv/get_int.hpp>
+#include <webots_ros2_msgs/srv/set_float.hpp>
+#include <webots_ros2_msgs/srv/set_int.hpp>
+#include <webots_ros2_msgs/srv/set_string.hpp>
+
+namespace webots_ros2_driver {
+  class Ros2Emitter : public Ros2SensorPlugin {
+  public:
+    // Your plugin has to override step() and init() methods
```
Comment (omichel) [inline webots_ros2_driver/include/webots_ros2_driver/plugins/static/Ros2Emitter.hpp:16]:
```suggestion
```

File: webots_ros2_driver/include/webots_ros2_driver/plugins/static/Ros2Receiver.hpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#include <webots_ros2_msgs/msg/string_stamped.hpp>
+#include <webots_ros2_msgs/srv/get_float.hpp>
+#include <webots_ros2_msgs/srv/get_int.hpp>
+#include <webots_ros2_msgs/srv/receiver_get_emitter_direction.hpp>
+#include <webots_ros2_msgs/srv/set_int.hpp>
+
+namespace webots_ros2_driver {
+  class Ros2Receiver : public Ros2SensorPlugin {
+  public:
+    // Your plugin has to override step() and init() methods
```
Comment (omichel) [inline webots_ros2_driver/include/webots_ros2_driver/plugins/static/Ros2Receiver.hpp:16]:
```suggestion
```

File: webots_ros2_driver/src/plugins/static/Ros2Emitter.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      mTopicName + "/set_range", std::bind(&Ros2Emitter::set_range_callback, this, _1, _2));
+    RCLCPP_DEBUG(rclcpp::get_logger(mDeviceName), "Emitter initialized!");
+  }
+  void Ros2Emitter::get_buffer_size_callback(const std::shared_ptr<webots_ros2_msgs::srv::GetInt::Request> request,
+                                             std::shared_ptr<webots_ros2_msgs::srv::GetInt::Response> response) {
+    int value = -1;
+    if (request->ask)
+      value = wb_emitter_get_buffer_size(mEmitter);
+
+    response->value = value;
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Emitter.cpp:43]:
```suggestion
    response->value = request->ask ? wb_emitter_get_buffer_size(mEmitter) : -1;
```

File: webots_ros2_driver/src/plugins/static/Ros2Emitter.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  }
+  void Ros2Emitter::get_channel_callback(const std::shared_ptr<webots_ros2_msgs::srv::GetInt::Request> request,
+                                         std::shared_ptr<webots_ros2_msgs::srv::GetInt::Response> response) {
+    int value = -1;
+    if (request->ask)
+      value = wb_emitter_get_channel(mEmitter);
+
+    response->value = value;
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Emitter.cpp:51]:
```suggestion
    response->value = request->ask ? wb_emitter_get_channel(mEmitter) : -1; 
```

File: webots_ros2_driver/src/plugins/static/Ros2Emitter.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                                  std::shared_ptr<webots_ros2_msgs::srv::SetString::Response> response) {
+    std::string message = request->value;
+    auto umessage = reinterpret_cast<unsigned char *>(const_cast<char *>(message.c_str()));
+    int ok = wb_emitter_send(mEmitter, umessage, message.length() + 1);
+    response->success = ok;
+  }
+  void Ros2Emitter::set_channel_callback(const std::shared_ptr<webots_ros2_msgs::srv::SetInt::Request> request,
+                                         std::shared_ptr<webots_ros2_msgs::srv::SetInt::Response> response) {
+    int channel = request->value;
+    wb_emitter_set_channel(mEmitter, channel);
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Emitter.cpp:71]:
```suggestion
    wb_emitter_set_channel(mEmitter, request->value);
```

File: webots_ros2_driver/src/plugins/static/Ros2Emitter.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    response->success = true;
+  }
+  void Ros2Emitter::set_range_callback(const std::shared_ptr<webots_ros2_msgs::srv::SetFloat::Request> request,
+                                       std::shared_ptr<webots_ros2_msgs::srv::SetFloat::Response> response) {
+    double range = request->value;
+    wb_emitter_set_range(mEmitter, range);
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Emitter.cpp:77]:
```suggestion
    wb_emitter_set_range(mEmitter, request->value);
```

File: webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    }
+    return mIsEnabled;
+  }
+  void Ros2Receiver::publishData() {
+    // If there is any packet publish the data
+    if (wb_receiver_get_queue_length(mReceiver) > 0) {
+      const int length_buffer = wb_receiver_get_data_size(mReceiver);
+      const void *received_data = wb_receiver_get_data(mReceiver);
+      // cast to char*
+      char *publishing_data_char = (char *)realloc(const_cast<void *>(received_data), length_buffer);
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp:75]: Why do we need to realloc here? That seems useless. It seems we only need to perform a cast, isn't it?

File: webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      // publish
+      mDataMessage.data = std::string(publishing_data_char);
+      mDataMessage.header.stamp = mNode->get_clock()->now();
+      mDataPublisher->publish(mDataMessage);
+      // release pointer
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp:80]: There is no release here...
```suggestion
```

File: webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+      wb_receiver_next_packet(mReceiver);
+    }
+  }
+  void Ros2Receiver::enable_callback(const std::shared_ptr<webots_ros2_msgs::srv::SetInt::Request> request,
+                                     std::shared_ptr<webots_ros2_msgs::srv::SetInt::Response> response) {
+    bool must_activate = request->value;
+    if (must_activate) {
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp:87]:
```suggestion
    if (request->value) {
```

File: webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+      wb_receiver_enable(mReceiver, mPublishTimestepSyncedMs);
+      mAlwaysOn = true;
+      mIsEnabled = true;
+    }
+    else {
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp:92]:
```suggestion
    } else {
```

File: webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+    response->direction = value;
+  }
+  void Ros2Receiver::get_sampling_period_callback(const std::shared_ptr<webots_ros2_msgs::srv::GetInt::Request> request,
+                                                  std::shared_ptr<webots_ros2_msgs::srv::GetInt::Response> response) {
+    int value = 0;
+    if (request->ask)
+      value = wb_receiver_get_sampling_period(mReceiver);
+
+    response->value = value;
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp:118]:
```suggestion
    response->value = request->ask ? wb_receiver_get_sampling_period(mReceiver) : 0;
```

File: webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+  }
+  void Ros2Receiver::get_signal_strength_callback(const std::shared_ptr<webots_ros2_msgs::srv::GetFloat::Request> request,
+                                                  std::shared_ptr<webots_ros2_msgs::srv::GetFloat::Response> response) {
+    double value = 0;
+    if (request->ask)
+      value = wb_receiver_get_signal_strength(mReceiver);
+
+    response->value = value;
```
Comment (omichel) [inline webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp:126]:
```suggestion
    response->value = request->ask ? wb_receiver_get_signal_strength(mReceiver) : 0;
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Comment by omichel** [comment]
There are still several clang format issues.

File: webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp (hunk 7 above)
Comment (angel-ayala) [reply webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp:75]: Yes, you are right, there were previous tests that troubled the decoding part but that was only an wrong length value on the `wb_emitted_send()` function from `Ros2Emitter`.

```suggestion
      char *publishing_data_char = static_cast<void *>(const_cast<void *>(received_data));
```

File: webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+      else
+        wb_receiver_disable(mReceiver);
+      mIsEnabled = shouldBeEnabled;
+    }
+    return mIsEnabled;
+  }
+  void Ros2Receiver::publishData() {
+    // If there is any packet publish the data
+    if (wb_receiver_get_queue_length(mReceiver) > 0) {
+      const int length_buffer = wb_receiver_get_data_size(mReceiver);
```
Comment (angel-ayala) [inline webots_ros2_driver/src/plugins/static/Ros2Receiver.cpp:72]:
```suggestion
```

**Comment by angel-ayala** [comment]
> There are still several clang format issues.

Exists a method to run the ClangFormat compliance before push the commit?
I'm executing the `test_source.py` file inside the tests folder and is not showing that output.

**Review by omichel — APPROVED** [review]
Thank you!
You can run clang-format locally and even better, integrate it within your favorite source code editor so that when you save, it will automatically format the C++ file using clang-format, so that you don't have to worry about it. Some instructions are provided [here](https://github.com/cyberbotics/webots/wiki/CPP-Coding-Style/#cs300-cyberbotics-clang-format-style-compliance) but you may find more up-to-date instructions elsewhere for your favorite source code editor.

**Pull request merged**